File: backend/app/routers/services.py
"""
Services Router — list services, subscribe/unsubscribe.
"""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.deps import get_current_user, get_optional_user
from app.models import Service, Subscription, User
from app.schemas import ServiceList, ServiceOut, SubscriptionCreate, SubscriptionOut
from app.tiers import can_subscribe_more, get_limits, get_upgrade_reason

router = APIRouter(prefix="/services", tags=["services"])


@router.get("", response_model=ServiceList)
async def list_services(
    category: str | None = None,
    search: str | None = None,
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    user: User | None = Depends(get_optional_user),
):
    """List all monitored services with optional filtering."""
    query = select(Service).where(Service.is_active == True)

    if category:
        query = query.where(Service.category == category)
    if search:
        query = query.where(Service.name.ilike(f"%{search}%"))

    # Count total
    count_q = select(func.count()).select_from(query.subquery())
    total = (await db.execute(count_q)).scalar() or 0

    # Paginate
    query = query.order_by(Service.name).offset((page - 1) * per_page).limit(per_page)
    result = await db.execute(query)
    services = result.scalars().all()

    # Get subscriber counts
    sub_counts = {}
    if services:
        svc_ids = [s.id for s in services]
        count_result = await db.execute(
            select(Subscription.service_id, func.count(Subscription.id))
            .where(Subscription.service_id.in_(svc_ids))
            .group_by(Subscription.service_id)
        )
        sub_counts = dict(count_result.all())

    service_list = []
    for svc in services:
        out = ServiceOut.model_validate(svc)
        out.subscriber_count = sub_counts.get(svc.id, 0)
        service_list.append(out)

    return ServiceList(services=service_list, total=total)


@router.get("/{slug}", response_model=ServiceOut)
async def get_service(
    slug: str,
    db: AsyncSession = Depends(get_db),
):
    """Get a single service by slug."""
    result = await db.execute(select(Service).where(Service.slug == slug))
    service = result.scalar_one_or_none()

    if not service:
        raise HTTPException(status_code=404, detail="Service not found")

    # Get subscriber count
    count_result = await db.execute(
        select(func.count(Subscription.id)).where(
            Subscription.service_id == service.id
        )
    )
    out = ServiceOut.model_validate(service)
    out.subscriber_count = count_result.scalar() or 0

    return out


@router.get("/categories/list")
async def list_categories(db: AsyncSession = Depends(get_db)):
    """Get all available service categories."""
    result = await db.execute(
        select(Service.category, func.count(Service.id))
        .where(Service.is_active == True, Service.category.isnot(None))
        .group_by(Service.category)
        .order_by(Service.category)
    )
    return [{"name": row[0], "count": row[1]} for row in result.all()]


# ── Subscriptions ────────────────────────────────────────────


@router.post("/subscribe", response_model=SubscriptionOut, status_code=201)
async def subscribe(
    body: SubscriptionCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Subscribe to change alerts for a service."""
    # ── Tier gate: check subscription limit ─────────────
    current_count = (await db.execute(
        select(func.count(Subscription.id)).where(Subscription.user_id == user.id)
    )).scalar() or 0

    if not can_subscribe_more(user.plan, current_count):
        limits = get_limits(user.plan)
        raise HTTPException(
            status_code=403,
            detail=(
                f"Your {user.plan.value} plan allows {limits.max_services} subscriptions. "
                f"Upgrade to Pro ($19/mo) for up to 15 services."
            ),
        )

    # ── Tier gate: webhook access ────────────────────────
    if body.notify_webhook:
        reason = get_upgrade_reason(user.plan, "webhook")
        if reason:
            raise HTTPException(status_code=403, detail=reason)

    # Check service exists
    svc = await db.execute(select(Service).where(Service.id == body.service_id))
    service = svc.scalar_one_or_none()
    if not service:
        raise HTTPException(status_code=404, detail="Service not found")

    # Check existing subscription
    existing = await db.execute(
        select(Subscription).where(
            Subscription.user_id == user.id,
            Subscription.service_id == body.service_id,
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(status_code=409, detail="Already subscribed")

    sub = Subscription(
        user_id=user.id,
        service_id=body.service_id,
        notify_email=body.notify_email,
        notify_webhook=body.notify_webhook,
        webhook_url=body.webhook_url,
    )
    db.add(sub)
    await db.flush()

    out = SubscriptionOut.model_validate(sub)
    out.service_name = service.name
    return out


@router.delete("/unsubscribe/{service_id}", status_code=204)
async def unsubscribe(
    service_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Unsubscribe from a service."""
    result = await db.execute(
        select(Subscription).where(
            Subscription.user_id == user.id,
            Subscription.service_id == service_id,
        )
    )
    sub = result.scalar_one_or_none()
    if sub:
        await db.delete(sub)


@router.get("/me/subscriptions", response_model=list[SubscriptionOut])
async def my_subscriptions(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get current user's subscriptions."""
    result = await db.execute(
        select(Subscription, Service.name)
        .join(Service, Subscription.service_id == Service.id)
        .where(Subscription.user_id == user.id)
        .order_by(Service.name)
    )
    subs = []
    for sub, svc_name in result.all():
        out = SubscriptionOut.model_validate(sub)
        out.service_name = svc_name
        subs.append(out)
    return subs

File: worker/src/routes/services.ts
import { Hono } from "hono";
import { eq, desc, and, sql, count as countFn } from "drizzle-orm";
import { getDb } from "../db/client";
import { services, subscriptions } from "../db/schema";
import { authOptional, authRequired, type AuthUser } from "../middleware/auth";
import type { Env, Plan } from "../lib/types";
import { TIER_LIMITS } from "../lib/types";

const app = new Hono<{ Bindings: Env }>();

/** GET /api/services — list all services */
app.get("/", authOptional, async (c) => {
  const db = getDb(c.env.DATABASE_URL);
  const page = Math.max(parseInt(c.req.query("page") ?? "1"), 1);
  const perPage = Math.min(parseInt(c.req.query("per_page") ?? "50"), 100);
  const category = c.req.query("category");

  let conditions: any[] = [eq(services.isActive, true)];
  if (category) conditions.push(eq(services.category, category));
  const where = conditions.length === 1 ? conditions[0] : and(...conditions);

  const rows = await db.select({
    id: services.id,
    name: services.name,
    slug: services.slug,
    category: services.category,
    logoUrl: services.logoUrl,
    website: services.website,
    tosUrl: services.tosUrl,
    privacyUrl: services.privacyUrl,
    lastCheckedAt: services.lastCheckedAt,
  })
    .from(services)
    .where(where)
    .orderBy(services.name)
    .limit(perPage)
    .offset((page - 1) * perPage);

  // Get subscriber counts
  const subCounts = await db.select({
    serviceId: subscriptions.serviceId,
    count: countFn(),
  })
    .from(subscriptions)
    .groupBy(subscriptions.serviceId);

  const countMap = new Map(subCounts.map((s) => [s.serviceId, s.count]));

  return c.json({
    items: rows.map((r) => ({
      id: r.id,
      name: r.name,
      slug: r.slug,
      category: r.category,
      logo_url: r.logoUrl,
      website: r.website,
      tos_url: r.tosUrl,
      privacy_url: r.privacyUrl,
      last_checked_at: r.lastCheckedAt,
      subscriber_count: countMap.get(r.id) ?? 0,
    })),
    page,
    per_page: perPage,
  });
});

/** GET /api/services/categories/list */
app.get("/categories/list", async (c) => {
  const db = getDb(c.env.DATABASE_URL);
  const rows = await db.select({
    category: services.category,
    count: countFn(),
  })
    .from(services)
    .where(eq(services.isActive, true))
    .groupBy(services.category)
    .orderBy(services.category);

  return c.json(rows.map((r) => ({
    category: r.category,
    count: r.count,
  })));
});

/** GET /api/services/:slug */
app.get("/:slug", async (c) => {
  const slug = c.req.param("slug");
  if (slug === "me" || slug === "categories") return; // let other routes handle

  const db = getDb(c.env.DATABASE_URL);
  const rows = await db.select().from(services).where(eq(services.slug, slug)).limit(1);
  if (rows.length === 0) {
    return c.json({ detail: "Service not found" }, 404);
  }
  const s = rows[0];
  return c.json({
    id: s.id,
    name: s.name,
    slug: s.slug,
    category: s.category,
    logo_url: s.logoUrl,
    website: s.website,
    tos_url: s.tosUrl,
    privacy_url: s.privacyUrl,
    last_checked_at: s.lastCheckedAt,
  });
});

/** POST /api/services/subscribe */
app.post("/subscribe", authRequired, async (c) => {
  const db = getDb(c.env.DATABASE_URL);
  const user = c.get("user") as AuthUser;
  const limits = TIER_LIMITS[user.plan as Plan];

  const body = await c.req.json<{ service_id: string; webhook_url?: string }>();

  // Check service exists
  const svc = await db.select({ id: services.id })
    .from(services).where(eq(services.id, body.service_id)).limit(1);
  if (svc.length === 0) {
    return c.json({ detail: "Service not found" }, 404);
  }

  // Check subscription limit
  const [subCount] = await db.select({ count: countFn() })
    .from(subscriptions).where(eq(subscriptions.userId, user.id));
  if ((subCount.count as number) >= limits.maxServices) {
    return c.json({ detail: `Your plan allows max ${limits.maxServices} subscriptions. Upgrade for more.` }, 403);
  }

  // Check if already subscribed
  const existing = await db.select({ id: subscriptions.id })
    .from(subscriptions)
    .where(and(eq(subscriptions.userId, user.id), eq(subscriptions.serviceId, body.service_id)))
    .limit(1);
  if (existing.length > 0) {
    return c.json({ detail: "Already subscribed" }, 409);
  }

  const wantWebhook = !!body.webhook_url;
  if (wantWebhook && !limits.webhookAccess) {
    return c.json({ detail: "Webhook alerts require Business plan" }, 403);
  }

  const [sub] = await db.insert(subscriptions).values({
    id: crypto.randomUUID(),
    userId: user.id,
    serviceId: body.service_id,
    notifyEmail: true,
    notifyWebhook: wantWebhook,
    webhookUrl: body.webhook_url ?? null,
    createdAt: new Date(),
  }).returning();

  return c.json({ id: sub.id, message: "Subscribed" });
});

/** DELETE /api/services/unsubscribe/:id */
app.delete("/unsubscribe/:id", authRequired, async (c) => {
  const db = getDb(c.env.DATABASE_URL);
  const user = c.get("user") as AuthUser;
  const serviceId = c.req.param("id");

  const deleted = await db.delete(subscriptions)
    .where(and(eq(subscriptions.userId, user.id), eq(subscriptions.serviceId, serviceId)))
    .returning({ id: subscriptions.id });

  if (deleted.length === 0) {
    return c.json({ detail: "Subscription not found" }, 404);
  }
  return c.json({ message: "Unsubscribed" });
});

/** GET /api/services/me/subscriptions */
app.get("/me/subscriptions", authRequired, async (c) => {
  const db = getDb(c.env.DATABASE_URL);
  const user = c.get("user") as AuthUser;

  const rows = await db.select({
    id: subscriptions.id,
    serviceId: subscriptions.serviceId,
    serviceName: services.name,
    serviceSlug: services.slug,
    notifyEmail: subscriptions.notifyEmail,
    notifyWebhook: subscriptions.notifyWebhook,
    createdAt: subscriptions.createdAt,
  })
    .from(subscriptions)
    .innerJoin(services, eq(subscriptions.serviceId, services.id))
    .where(eq(subscriptions.userId, user.id))
    .orderBy(services.name);

  return c.json(rows.map((r) => ({
    id: r.id,
    service_id: r.serviceId,
    service_name: r.serviceName,
    service_slug: r.serviceSlug,
    notify_email: r.notifyEmail,
    notify_webhook: r.notifyWebhook,
    created_at: r.createdAt,
  })));
});

export default app;
